### ISSUE TITLE
Where can decrypt the prediction results? And who is responsible for decrypting them?

### ISSUE THREAD

**Issue by qxzhou1010** [zama-ai/concrete-ml#545]
In some of the demos provided in the documentation, I noticed that predictions under encryption only require one line of code, and the prediction results are directly in plaintext.  Such as: `y_pred_fhe = model.predict(X_test, fhe="execute")`.

This raises the question, **where does the decryption take place? Is it decrypted on the client-side or the server-side?**

Moreover, in step-by-step examples, it seems that as long as one holds the `model` object, one can complete all functions including encryption, encrypted prediction, and decryption. However, in practical use, cloud servers often only need to perform encrypted predictions, obtain encrypted results, and send them back to the client, who then decrypts the results locally to obtain the plaintext.

```
...
q_input_enc = model.fhe_circuit.encrypt(q_input)
...
q_y_enc = model.fhe_circuit.run(q_input_enc)
....
q_y = model.fhe_circuit.decrypt(q_y_enc)
```

Therefore, if I want to perform the following steps using concrete-ml, what should I do?

1. The client generates a public key and a private key.
2. The client encrypts their own data and model, then sends the encrypted model and data to the cloud server.
3. The cloud server completes the model prediction in an encrypted state and returns the encrypted prediction results to the client.
4. The client decrypts the prediction results.

**Comment by fd0r** [comment]
Hello @qxzhou1010 , very good question!

We provide the `fhe="execute"` option mostly so that developers can check that the predictions in FHE will be the same as in the clear and to validate that the FHE runtime is acceptable for the given use-case.

For deployment purposes I invite you to check https://docs.zama.ai/concrete-ml/deployment/client_server and the associated notebooks linked in the page.

Encryption and decryption should always happen on the client side.

> Therefore, if I want to perform the following steps using concrete-ml, what should I do?

I hope the following notebook https://github.com/zama-ai/concrete-ml/blob/release/1.4.x/docs/advanced_examples/ClientServer.ipynb will answer this question but don't hesitate to follow-up here if you have any doubt about it.

**Comment by qxzhou1010** [comment]
Thank you for your reply; I no longer have any issues on this point.

I have read the article about [inference in the cloud](https://docs.zama.ai/concrete-ml/getting-started/cloud). I have two new questions:

1. As mentioned in [issue #546](https://github.com/zama-ai/concrete-ml/issues/546), the compiled model currently exists in plaintext form. In the system setup for [cloud inference](https://docs.zama.ai/concrete-ml/getting-started/cloud), the model developer completes the development and compilation of the model locally and then sends it to the server. **Does this mean that the model's parameters are potentially exposed to the server? In this situation, would model developers be willing to deploy their trained models on cloud servers?**
2. My second question concerns the client side. In the system setup for [cloud inference](https://docs.zama.ai/concrete-ml/getting-started/cloud), the client initially requests encryption parameters from the server and then generates a private key and evaluation key locally. My question is about multi-key FHE . In this scenario, **can a single server provide model prediction services to multiple clients simultaneously?** That is, if different clients initiate requests to the server, are the keys generated by these clients supposed to be different? **If so, can it be understood that the underlying technology uses multi-key FHE techniques?**

**Comment by fd0r** [comment]
> Does this mean that the model's parameters are potentially exposed to the server? In this situation, would model developers be willing to deploy their trained models on cloud servers?

Yes the model's parameters will be exposed to the server.
If that's an issue or not will depend on each use-case. 
One can always self-host the FHE models if using a cloud provider is an issue for them.

> In this scenario, can a single server provide model prediction services to multiple clients simultaneously? That is, if different clients initiate requests to the server, are the keys generated by these clients supposed to be different? If so, can it be understood that the underlying technology uses multi-key FHE techniques?

I don't know much about multi-key FHE techniques to be honest but I'm fairly confident that this isn't implemented in Concrete at the moment.
Multiple clients will require multiple keys. And yes of course a single server can serve multiple users, it will just need to load the proper evaluation key when a user makes a request.

Basically it goes like:
1. The client requests the crypto-parameters from the server to properly generate a private key and evaluation key.
2. The client sends its evaluation key to the server that stores it with some identifier that is sent to the client.
3. The client encrypts data with its private key and sends the encrypted data to the server for computation with its user identifier.
4. The server loads the appropriate evaluation key to do the computation, does the computation and sends the encrypted result back to the user
5. The user decrypts the result using its private key

**Comment by qxzhou1010** [comment]
Thank you very much for your so quickly response.

Regarding question 2, I have understood it now. **Since the model is in plaintext, it can naturally support different clients.** The clients simply need to ensure that their input data is encrypted. Thus, the plaintext model combined with encrypted input to produce an encrypted output will naturally use keys generated by the clients themselves.
